hazelcast/src/main/java/com/hazelcast/map/impl/operation/QueryOperation.java

```diff
@@ -114,6 +111,41 @@ public void run() throws Exception {
         checkPartitionStateChanges(partitionService, partitionStateVersion);
     }
 
+    private void fullTableScan(Collection<Integer> initialPartitions, GroupProperties groupProperties)
+            throws InterruptedException, ExecutionException {
+        if (pagingPredicate != null) {
+            runParallelForPaging(initialPartitions);
+        } else {
+            boolean parallelEvaluation = groupProperties.QUERY_PREDICATE_PARALLEL_EVALUATION.getBoolean();
+            if (parallelEvaluation) {
+                runParallel(initialPartitions);
+            } else {
+                runSingleThreaded(initialPartitions);
+            }
+        }
+    }
+
+    protected void runSingleThreaded(final Collection<Integer> initialPartitions) {
+        RuntimeException storedException = null;
+        MapServiceContext mapServiceContext = mapService.getMapServiceContext();
+        MapContextQuerySupport querySupport = mapServiceContext.getMapContextQuerySupport();
+        for (Integer partitionId : initialPartitions) {
+            try {
+                Collection<QueryableEntry> entries = querySupport.queryOnPartition(name, predicate, partitionId);
+                if (!addQueryEntriesToResult(entries)) {
+                    break;
+                }
+            } catch (RuntimeException e) {
+                if (storedException == null) {
```

Comment:
Yes, thanks! I still have to look into the (slow) MapLoader thing, but your explanation was very helpful.

So I have no objection to the exception approach of your method anymore :)